File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.cc
#include<iostream>

using namespace std;

int N;
int arr[9] = { 0, };
int visit[9] = { 0, };
int sol[9] = { 0, };
int s = 0;

void dfs(int count) {
	if (count == N) {
		int tmp = 0;

		for (int i = 1; i < N; i++) {
			tmp += abs(sol[i] - sol[i + 1]);
		}

		if (s < tmp) {
			s = tmp;
		}
	}

	for (int i = 1; i <= N; i++) {

		if (visit[i] != 1) {
			sol[count+1] = arr[i];
			visit[i] = 1;
			dfs(count + 1);
			visit[i] = 0;
		}
	}

}

int main() {


	cin >> N;

	for (int i = 1; i <= N; i++) {
		cin >> arr[i];
	}

	dfs(0);

	cout << s;

	return 0;
}

File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.java
import java.io.*;

public class Main{
    public static int[] check;
    public static int[] sol;
    public static int answer = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int N = Integer.parseInt(br.readLine());
        String[] str = br.readLine().split(" ");
        int[] arr  = new int[N];
        check = new int[N];
        sol = new int[N];

        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(str[i]);
        }

        rec(arr,N,0);

        bw.write(answer+"");
        bw.flush();
        bw.close();

    }

    public static void rec(int[] arr, int n, int count) {
        if (n == count) {
            int tmp = 0;
            for (int i = 1; i < sol.length; i++) {
                if (sol[i] - sol[i - 1] < 0) {
                    tmp += sol[i-1] - sol[i];
                }
                else {
                    tmp += sol[i] - sol[i-1];
                }
            }
//            System.out.println("수 배열 완성! total= " + tmp);
            if (tmp > answer) {
                answer = tmp;
            }
            return;
        }

        for (int i = 0; i < arr.length; i++) {
            if (check[i] == 0) {
                check[i] = 1;
                sol[count] = arr[i];
                rec(arr, n, count + 1);
                check[i] = 0;
            }
        }
    }
}